sprandom.c

```diff
@@ -375,9 +381,197 @@ int sprandom_setup(struct sprandom_info *spr_info, uint64_t logical_size, uint64
 	total_alloc += spr_info->num_regions * sizeof(spr_info->validity_dist[0]);
 
 	/* Precompute invalidity percentage array */
+	spr_info->invalid_pct = calloc(spr_info->num_regions,
+				       sizeof(spr_info->invalid_pct[0]));
+	if (!spr_info->invalid_pct)
+		goto err;
+
+	total_alloc += spr_info->num_regions * sizeof(spr_info->invalid_pct[0]);
+
+	for (i = 0; i < spr_info->num_regions; i++) {
+		double inv = (1.0 - validity_dist[i]) * (double)PCT_PRECISION;
+		spr_info->invalid_pct[i] = (int)round(inv);
+	}
+
 	region_sz = physical_size / spr_info->num_regions;
+	region_write_count = region_sz / align_bs;
+
+	invalid_capacity = 2 * ceil((double)region_write_count * (1.0 - validity_dist[0]));
+
+	spr_info->invalid_buf = pcb_alloc(invalid_capacity);
+
+	total_alloc += invalid_capacity * sizeof(uint64_t);
 
 	spr_info->region_sz = region_sz;
+	spr_info->invalid_count[0] = 0;
+	spr_info->invalid_count[1] = 0;
+	spr_info->curr_phase = 0;
+	spr_info->current_region = 0;
+	spr_info->writes_remaining = region_write_count;
+	spr_info->region_write_count = region_write_count;
+
+
+	/* Display overall allocation */
+	dprint(FD_SPRANDOM, "Overall allocation:\n");
+	dprint(FD_SPRANDOM, "  logical_size:      %lu: %s\n",
+		logical_size,
+		bytes2str_simple(bytes2str_buf, sizeof(bytes2str_buf), logical_size));
+	dprint(FD_SPRANDOM, "  physical_size:     %lu: %s\n",
+		physical_size,
+		bytes2str_simple(bytes2str_buf, sizeof(bytes2str_buf), physical_size));
+	dprint(FD_SPRANDOM, "  region_size:       %lu\n", region_sz);
+	dprint(FD_SPRANDOM, "  num_regions:       %u\n", spr_info->num_regions);
+	dprint(FD_SPRANDOM, "  region_write_count:%lu\n", region_write_count);
+	dprint(FD_SPRANDOM, "  invalid_capacity:  %lu\n", invalid_capacity);
+	dprint(FD_SPRANDOM, "  dynamic memory:    %zu: %s\n",
+		total_alloc,
+		bytes2str_simple(bytes2str_buf, sizeof(bytes2str_buf), total_alloc));
 
 	return 0;
+err:
+	free(spr_info->validity_dist);
+	free(spr_info->invalid_pct);
+	return -ENOMEM;
+}
+
+static void sprandom_add_with_probability(struct sprandom_info *info,
+					  uint64_t offset, unsigned int phase)
+{
+	int v = rand_between(&info->rand_state, 0, PCT_PRECISION);
+
+	if (v <= info->invalid_pct[info->current_region]) {
+		if (pcb_space_available(info->invalid_buf)) {
+			pcb_push_staged(info->invalid_buf, offset);
+			info->invalid_count[phase]++;
+		} else {
+			dprint(FD_SPRANDOM, "pcb buffer full!!!\n");
+			assert(false);
+		}
+	}
+}
+
+static void dprint_invalidation(const struct sprandom_info *info)
+{
+	uint32_t phase = info->curr_phase;
+
+	dprint(FD_SPRANDOM, "Invalidation[%d] %ld %ld %.04f %d\n",
+		info->current_region,
+		info->region_write_count,
+		info->invalid_count[phase],
+		(double)info->invalid_count[phase] / (double)info->region_write_count,
+		info->invalid_pct[info->current_region]);
+
+}
+
+int sprandom_get_next_offset(struct sprandom_info *info, struct fio_file *f, uint64_t *b)
```

Comment:
I think this function deserves a summary explanation at the top. Something along the lines of:
- This generates offsets to write a region and saves a fraction of the offsets in a ring buffer
- The next region starts by writing all of the saved offsets to achieve the desired fraction of invalid blocks in the previous region. Then the remainder of the current region is filled with new offsets.